### PR TITLE
Enhanced CLI stats

### LINKED ISSUE #1595: Display more encoding stats at end of encode
Tracking this for purposes of collaboration, although I plan to work on it this week. Looking to implement stats similar to what x264 shows at end of encode. This could avoid having to go into aom-analyzer for things rav1e could easily count for us, e.g. if no skip blocks are being used.

```
y4m [info]: 640x360p 1:1 @ 50/1 fps (cfr)
x264 [info]: using SAR=1/1
x264 [info]: using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x264 [info]: profile High, level 3.1
x264 [info]: frame I:1     Avg QP:30.55  size: 29856
x264 [info]: frame P:3     Avg QP:32.69  size: 11936
x264 [info]: frame B:6     Avg QP:33.68  size:   869
x264 [info]: consecutive B-frames: 20.0%  0.0%  0.0% 80.0%
x264 [info]: mb I  I16..4:  0.0% 71.5% 28.5%
x264 [info]: mb P  I16..4:  0.1%  2.4%  0.3%  P16..4: 35.3% 31.0% 15.7%  0.0%  0.0%    skip:15.2%
x264 [info]: mb B  I16..4:  0.0%  0.0%  0.0%  B16..8: 25.2%  0.6%  0.2%  direct: 3.2%  skip:70.8%  L0:13.0% L1:42.7% BI:44.3%
x264 [info]: 8x8 transform intra:72.6% inter:70.7%
x264 [info]: coded y,uvDC,uvAC intra: 97.9% 96.8% 88.3% inter: 26.2% 13.2% 4.0%
x264 [info]: i16 v,h,dc,p:  0%  0%  0% 100%
x264 [info]: i8 v,h,dc,ddl,ddr,vr,hd,vl,hu:  4% 53%  7%  2%  2%  1% 17%  1% 13%
x264 [info]: i4 v,h,dc,ddl,ddr,vr,hd,vl,hu:  3% 53%  4%  3%  4%  2% 18%  1% 13%
x264 [info]: i8c dc,h,v,p: 49% 38%  6%  7%
x264 [info]: Weighted P-Frames: Y:0.0% UV:0.0%
x264 [info]: ref P L0: 83.4% 12.9%  3.0%  0.7%
x264 [info]: kb/s:2835.20

encoded 10 frames, 126.63 fps, 2835.20 kb/s
```

### PR DESCRIPTION
**Pull request by shssoichiro**
Closes #1595

Example of verbose output:
```
[INFO  rav1e] Using y4m decoder: 1280x720p @ 60/1 fps, 4:2:0, 8-bit
[INFO  rav1e] CPU Feature Level: AVX2
[INFO  rav1e] Encoding settings: keyint_min=12 keyint_max=240 quantizer=100 bitrate=0 min_quantizer=0 low_latency=false tune=Psychovisual tiles=0 tile_rows=0 tile_cols=0 rdo_lookahead_frames=40 min_block_size=4x4 multiref=true fast_deblock=false reduced_tx_set=false tx_domain_distortion=true tx_domain_rate=false encode_bottomup=true rdo_tx_decision=true prediction_modes=Complex-All include_near_mvs=true no_scene_detection=false diamond_me=true cdef=true quantizer_rdo=true use_satd_subpel=true
[INFO  rav1e] Input Frame 0 - Key frame - 59872 bytes - encoded 1 frames, 0.003 fps, 28738.56 Kb/s
[INFO  rav1e] Input Frame 1 - Inter frame - 27367 bytes - encoded 2 frames, 0.003 fps, 20937.36 Kb/s
...
[INFO  rav1e] Input Frame 59 - Inter frame - 2068 bytes - encoded 60 frames, 0.011 fps, 3265.33 Kb/s
[INFO  rav1e::stats] ----------
[INFO  rav1e::stats] Key frame:             1 | avg QP:  79.00 | avg size:   59872 B
[INFO  rav1e::stats] Inter frame:          59 | avg QP: 117.29 | avg size:    5903 B
[INFO  rav1e::stats] Intra only frame:      0 | avg QP:   0.00 | avg size:       0 B
[INFO  rav1e::stats] Switching frame:       0 | avg QP:   0.00 | avg size:       0 B
[INFO  rav1e::stats] ----------
[INFO  rav1e::stats] bsize I:   x128    x64    x32    x16     x8     x4
[INFO  rav1e::stats]    128x:   0.0%   0.0%                              
[INFO  rav1e::stats]     64x:   0.0%   0.0%   0.4%   0.0%
[INFO  rav1e::stats]     32x:          0.3%   0.1%   1.5%   0.0%
[INFO  rav1e::stats]     16x:          0.0%   1.3%   1.1%   6.5%   0.0%
[INFO  rav1e::stats]      8x:                 0.0%   5.6%   6.8%  21.4%
[INFO  rav1e::stats]      4x:                        0.0%  20.8%  34.1%
[INFO  rav1e::stats] ----------
[INFO  rav1e::stats] bsize P:   x128    x64    x32    x16     x8     x4
[INFO  rav1e::stats]    128x:   0.0%   0.0%                              skip:  98.2%
[INFO  rav1e::stats]     64x:   0.0%   0.1%   0.7%   0.0%
[INFO  rav1e::stats]     32x:          0.7%   0.6%   2.7%   0.0%
[INFO  rav1e::stats]     16x:          0.0%   2.6%   3.2%   9.1%   0.0%
[INFO  rav1e::stats]      8x:                 0.0%   7.9%  12.6%  19.2%
[INFO  rav1e::stats]      4x:                        0.0%  16.3%  24.3%
[INFO  rav1e::stats] ----------
[INFO  rav1e::stats] txtypes I: DCT_DCT: 43.6% | ADST_DCT: 10.6% | DCT_ADST: 16.5% | ADST_ADST: 26.7%
[INFO  rav1e::stats]            IDTX: 0.3% | V_DCT: 0.5% | H_DCT: 1.7%
[INFO  rav1e::stats] txtypes P: DCT_DCT: 98.9% | ADST_DCT: 0.4% | DCT_ADST: 0.3% | ADST_ADST: 0.2%
[INFO  rav1e::stats]            IDTX: 0.0% | V_DCT: 0.0% | H_DCT: 0.1%
[INFO  rav1e::stats] ----------
[INFO  rav1e::stats] y modes I: DC: 18.3% | V: 3.0% | H: 20.1% | Paeth: 3.3%
[INFO  rav1e::stats]            Smooth: 25.0% | Smooth V: 6.7% | Smooth H: 5.8%
[INFO  rav1e::stats]         D: 45: 0.5% | 63: 1.7% | 117: 1.3% | 135: 2.1% | 153: 10.0% | 207: 2.3%
[INFO  rav1e::stats] uv modes I: DC: 23.8% | V: 2.7% | H: 19.0% | Paeth: 3.0%
[INFO  rav1e::stats]             Smooth: 21.7% | Smooth V: 5.4% | Smooth H: 4.3% | UV CFL: 4.3%
[INFO  rav1e::stats]          D: 45: 0.4% | 63: 1.4% | 117: 1.1% | 135: 1.8% | 153: 9.1% | 207: 2.0%
[INFO  rav1e::stats] ----------
[INFO  rav1e::stats] y modes P: Nearest: 64.0% | Near0: 12.7% | Near1: 0.1% | NearNear: 1.6%
[INFO  rav1e::stats]            New: 5.5% | NewNew: 0.8% | NearestNearest: 14.6% | GlobalGlobal: 0.0%
[INFO  rav1e::stats] uv modes P: Nearest: 64.0% | Near0: 12.7% | Near1: 0.1% | NearNear: 1.6%
[INFO  rav1e::stats]             New: 5.5% | NewNew: 0.8% | NearestNearest: 14.6% | GlobalGlobal: 0.0%
```

Non-verbose output:
```
[INFO  rav1e] Using y4m decoder: 640x360p @ 25/1 fps, 4:2:0, 8-bit
[INFO  rav1e] CPU Feature Level: AVX2
[INFO  rav1e] encoded 3/3 frames, 0.766 fps, 1886.33 Kb/s, est. size: 0.03 MB, est. time: 0 s      
[INFO  rav1e::stats] ----------
[INFO  rav1e::stats] Key frame:             1 | avg QP:  79.00 | avg size:   26893 B
[INFO  rav1e::stats] Inter frame:           2 | avg QP: 117.00 | avg size:     701 B
[INFO  rav1e::stats] Intra only frame:      0 | avg QP:   0.00 | avg size:       0 B
[INFO  rav1e::stats] Switching frame:       0 | avg QP:   0.00 | avg size:       0 B
```

### CODE REVIEW

File: src/bin/rav1e.rs [hunk 1]
```diff
@@ -266,20 +279,26 @@ fn run() -> Result<(), error::CliError> {
...
-    video_info.time_base.den as usize,
-    video_info.time_base.num as usize,
+    video_info.time_base.den,
+    video_info.time_base.num,
+    video_info.chroma_sampling,
+    video_info.bit_depth
   );
+  info!("CPU Feature Level: {}", cfg.enc.cpu_feature_level);
+  if cli.verbose {
+    info!("Encoding settings: {}", cfg.enc.print_settings());
```
Comment (shssoichiro) [inline src/bin/rav1e.rs]: Would it be useful to print this always? It's a long string, but the number of times someone has said "I didn't realize feature X isn't enabled at speed 5" is a lot.

File: src/api/mod.rs [hunk 2]
```diff
@@ -226,6 +227,45 @@ impl EncoderConfig {
   pub fn frame_rate(&self) -> f64 {
     Rational::from_reciprocal(self.time_base).as_f64()
   }
+
+  /// Print the current encoding settings into a string of key-value pairs
+  pub fn print_settings(&self) -> String {
```
Comment (YaLTeR) [inline src/api/mod.rs]: Maybe put this into `impl Display for EncoderSettings`?

File: src/bin/rav1e.rs [hunk 3]
```diff
@@ -206,22 +209,32 @@ fn do_encode<T: Pixel, D: Decoder>(
...
+        progress.add_frame(frame);
         eprint!("\r{}                    ", progress);
       };
     }
 
     output.flush().unwrap();
   }
-  eprint!("\n{}\n", progress.print_summary());
+  if !verbose {
+    eprint!("\r");
```
Comment (YaLTeR) [inline src/bin/rav1e.rs]: This should probably not be here

File: src/bin/rav1e.rs [hunk 4]
```diff
@@ -206,22 +209,32 @@ fn do_encode<T: Pixel, D: Decoder>(
...
+    info!("{}", progress);
+  }
+  progress.print_summary(verbose);
   Ok(())
 }
 
 fn main() {
   better_panic::install();
+  if env::var("RUST_LOG").is_err() {
+    env::set_var("RUST_LOG", "rav1e=info");
```
Comment (YaLTeR) [inline src/bin/rav1e.rs]: Not sure this is the correct way of doing this

File: src/encoder.rs [hunk 5]
```diff
@@ -944,6 +947,61 @@ impl<T: Pixel> fmt::Display for FrameInvariants<T> {
...
+      *new.tx_type_counts.entry(k).or_insert(0) += v;
+    });
+    new.skip_block_count += rhs.skip_block_count;
+    new
+  }
+}
+
+impl AddAssign<&Self> for EncoderStats {
+  fn add_assign(&mut self, rhs: &EncoderStats) {
+    let copy = self.clone();
```
Comment (YaLTeR) [inline src/encoder.rs]: I think this is usually done the other way around to avoid clones. You put logic in `AddAssign`, then in `Add` you do `self += rhs; self`

File: src/bin/rav1e.rs (hunk 3 above)
Comment (shssoichiro) [reply src/bin/rav1e.rs]: This is used to clear out the progress indicator at the end of encoding (in non-verbose mode). In verbose mode there is no progress indicator, we print every frame on a new line, so we don't need to clear anything on verbose mode.

File: src/bin/rav1e.rs (hunk 4 above)
Comment (shssoichiro) [reply src/bin/rav1e.rs]: You're right, looks like this is: https://docs.rs/env_logger/0.6.2/env_logger/#specifying-defaults-for-environment-variables

File: src/api/color.rs [hunk 6]
```diff
@@ -44,6 +45,21 @@ pub enum ChromaSampling {
   Cs400,
 }
 
+impl fmt::Display for ChromaSampling {
```
Comment (lu-zero) [inline src/api/color.rs]: This should be a different PR (and could land immediately)

File: src/transform/mod.rs [hunk 7]
```diff
@@ -41,7 +41,7 @@ static INV_SQRT2: i32 = 2896; // 2^12 / sqrt(2)
 
 pub const TX_TYPES: usize = 16;
 
-#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd, Eq, Ord)]
```
Comment (lu-zero) [inline src/transform/mod.rs]: This also should go in the other PR.

File: src/lib.rs [hunk 8]
```diff
@@ -46,6 +46,9 @@
 #[macro_use]
 extern crate pretty_assertions;
 
+#[macro_use]
+extern crate log;
```
Comment (lu-zero) [inline src/lib.rs]: The use of log could be another different PR.

File: src/encoder.rs [hunk 9]
```diff
@@ -2033,7 +2100,7 @@ pub fn encode_block_with_modes<T: Pixel>(
   fi: &FrameInvariants<T>, ts: &mut TileStateMut<'_, T>,
   cw: &mut ContextWriter, w_pre_cdef: &mut dyn Writer,
   w_post_cdef: &mut dyn Writer, bsize: BlockSize, tile_bo: TileBlockOffset,
-  mode_decision: &RDOPartitionOutput, rdo_type: RDOType,
+  mode_decision: &RDOPartitionOutput, rdo_type: RDOType, record_stats: bool,
```
Comment (lu-zero) [inline src/encoder.rs]: it should go in `Config`, not be a parameter.

File: src/encoder.rs [hunk 10]
```diff
@@ -1479,7 +1536,7 @@ pub fn encode_block_post_cdef<T: Pixel>(
   mvs: [MotionVector; 2], bsize: BlockSize, tile_bo: TileBlockOffset,
   skip: bool, cfl: CFLParams, tx_size: TxSize, tx_type: TxType,
   mode_context: usize, mv_stack: &[CandidateMV], rdo_type: RDOType,
-  need_recon_pixel: bool,
+  need_recon_pixel: bool, record_stats: bool,
```
Comment (lu-zero) [inline src/encoder.rs]: It should be checked as `fi.config.record_stats`

File: src/encoder.rs [hunk 11]
```diff
@@ -944,6 +947,60 @@ impl<T: Pixel> fmt::Display for FrameInvariants<T> {
   }
 }
 
+#[derive(Debug, Clone)]
```
Comment (lu-zero) [inline src/encoder.rs]: It should live in `stats.rs` or `encoder/stats.rs`

File: src/cpu_features.rs [hunk 12]
```diff
@@ -16,12 +16,27 @@ pub use x86::*;
...
 
+  use std::fmt;
+
   #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd)]
   pub enum CpuFeatureLevel {
     NATIVE,
     AVX2,
   }
 
+  impl fmt::Display for CpuFeatureLevel {
```
Comment (lu-zero) [inline src/cpu_features.rs]: `#[derive(ArgEnum)]` maybe?

File: src/api/mod.rs [hunk 13]
```diff
@@ -15,6 +15,7 @@ pub use color::*;
 
 use arrayvec::ArrayVec;
 use bitstream_io::*;
+use itertools::Itertools;
```
Comment (lu-zero) [inline src/api/mod.rs]: If toml-rs takes less time than itertools I'd consider using this + serde. (bonus point: being able to feed rav1e with the configuration used)

File: Cargo.toml [hunk 14]
```diff
@@ -14,7 +14,7 @@ autobins = false
 [features]
 decode_test = ["aom-sys"]
 decode_test_dav1d = ["dav1d-sys"]
-binaries = ["ivf", "y4m", "clap", "scan_fmt"]
+binaries = ["ivf", "y4m", "clap", "scan_fmt", "env_logger"]
```
Comment (lu-zero) [inline Cargo.toml]: You may use the pretty_env_logger instead.

File: src/encoder.rs (hunk 9 above)
Comment (shssoichiro) [reply src/encoder.rs]: I don't think it could, because it depends on where we're calling `encode_block` from. If we're calling from RDO analysis, we don't want to record stats because that's a trial encode. We only want to record stats for the final block encode.

File: src/bin/rav1e.rs (hunk 1 above)
Comment (lu-zero) [reply src/bin/rav1e.rs]: Probably it is a good idea.

File: src/encoder.rs (hunk 9 above)
Comment (lu-zero) [reply src/encoder.rs]: You are right.

File: src/api/color.rs (hunk 6 above)
Comment (shssoichiro) [reply src/api/color.rs]: I'm not sure there's a use in PR'ing this separately, because it doesn't do anything without being a part of this PR.

File: src/lib.rs (hunk 8 above)
Comment (shssoichiro) [reply src/lib.rs]: This one is reasonable, I can move the log migration to another PR.

File: src/api/mod.rs (hunk 13 above)
Comment (shssoichiro) [reply src/api/mod.rs]: I'm not sure if there's a clean way of doing this with toml-rs, given that this currently shows some of the fields from `EncoderConfig`, then all the fields in `SpeedSettings`. Unless you have any suggestions?

File: src/transform/mod.rs (hunk 7 above)
Comment (shssoichiro) [reply src/transform/mod.rs]: Similarly, this is only required to use this in a BTreeMap.

File: src/api/mod.rs (hunk 13 above)
Comment (lu-zero) [reply src/api/mod.rs]: I wonder why some encoder config fields are omitted. We can discuss further on IRC probably.

Comment (shssoichiro) [reply src/api/mod.rs]: I chose to hide some of the fields because I thought some of them didn't make sense to show, or we already showed them. But thinking about it more, the only ones that we already show are width, height, and bit_depth. These are ones that we wouldn't want to be user-settable, but I think we can configure toml-rs via serde to hide those fields and set them to a default value on load (so the encoder can set them properly).

Maybe it doesn't make sense for width, height, and bit_depth to be in EncoderConfig at all? That would be a larger discussion though.

Comment (shssoichiro) [reply src/api/mod.rs]: I took a quick stab at it and ended up getting a panic: `called `Result::unwrap()` on an `Err` value: ValueAfterTable` which I think is because we have nested structs in encoder config. It's probably possible to make it work, but I'm going to go ahead and land it as-is for now.

Comment (lu-zero) [reply src/api/mod.rs]: We can refine and improve later :)
